node/network/approval-distribution/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+			peers.clone(),
+			protocol_v1::ValidationProtocol::ApprovalDistribution(
+				protocol_v1::ApprovalDistributionMessage::Approvals(approvals)
+			),
+		).into()).await;
+
+		// Add the fingerprint of the assignment to the knowledge of each peer.
+		for peer in peers.into_iter() {
+			entry.known_by
+				.entry(peer)
+				.or_default()
+				.known_messages
+				.insert(fingerprint.clone());
+		}
+	}
+
+	async fn unify_with_peer(
+		entries: &mut HashMap<Hash, BlockEntry>,
+		ctx: &mut impl SubsystemContext<Message = ApprovalDistributionMessage>,
+		metrics: &Metrics,
+		peer_id: PeerId,
+		view: View,
+	) {
+		let mut to_send = HashSet::new();
+
+		let view_finalized_number = view.finalized_number;
+		for head in view.heads.into_iter() {
+			let mut block = head;
+			let interesting_blocks = std::iter::from_fn(|| {
+				// step 2.
+				let entry = match entries.get_mut(&block) {
+					Some(entry) if entry.number >= view_finalized_number => entry,
+					_ => return None,
+				};
+				let interesting_block = match entry.known_by.entry(peer_id.clone()) {
+					// step 3.
+					hash_map::Entry::Occupied(_) => return None,
+					// step 4.
+					hash_map::Entry::Vacant(vacant) => {
+						vacant.insert(entry.knowledge.clone());
+						block
+					}
+				};
+				// step 5.
+				block = entry.parent_hash.clone();
+				Some(interesting_block)
+			});
+			to_send.extend(interesting_blocks);
+		}
+		// step 6.
+		// send all assignments and approvals for all candidates in those blocks to the peer
+		Self::send_gossip_messages_to_peer(
+			entries,
+			ctx,
+			metrics,
+			peer_id,
+			to_send
+		).await;
+	}
+
+	#[tracing::instrument(level = "trace", skip(entries, ctx, _metrics, blocks), fields(subsystem = LOG_TARGET))]
+	async fn send_gossip_messages_to_peer(
+		entries: &HashMap<Hash, BlockEntry>,
+		ctx: &mut impl SubsystemContext<Message = ApprovalDistributionMessage>,
+		_metrics: &Metrics,
+		peer_id: PeerId,
+		blocks: HashSet<Hash>,
+	) {
+		let mut assignments = Vec::new();
+		let mut approvals = Vec::new();
+
+		for block in blocks.into_iter() {
+			let entry = match entries.get(&block) {
+				Some(entry) => entry,
+				None => continue, // should be unreachable
+			};
+			for (candidate_index, candidate_entry) in entry.candidates.iter() {
+				for (validator_index, approval_state) in candidate_entry.approvals.iter() {
+					match approval_state {
+						ApprovalState::Assigned(cert) => {
+							assignments.push((IndirectAssignmentCert {
+								block_hash: block.clone(),
+								validator: validator_index.clone(),
+								cert: cert.clone(),
+							}, candidate_index.clone()));
+						}
+						ApprovalState::Approved(_, signature) => {
+							approvals.push(IndirectSignedApprovalVote {
+								block_hash: block.clone(),
+								validator: validator_index.clone(),
+								candidate_index: candidate_index.clone(),
+								signature: signature.clone(),
+							});
+						}
+					}
+				}
+			}
+		}
+
+		if !assignments.is_empty() {
```

Comment:
Needs deduplication of assignments, as the same assignment may exist for multiple blocks.

Reply:
Do I understand correctly that I need to check for unique `AssignmentCert`s (sadly it doesn't impl `Hash` or `Ord`)? Because `IndirectAssignmentCert` contains a block_hash.

And no check for approvals is required.

Reply:
At this point you probably know this better than I do - I haven't really looked for a month. You're confident there are no duplicates?

Reply:
Is the argument that we don't need deduplication because the architecture of the subsystem actually intends that there should be a message for every block that a validator is assigned to a candidate via? That seems sound and yes I don't think we need deduplication if so.